README
Goo-Canvas version 0.04
=======================

Perl bindings to the goocanvas-0.9. For goocanvas-0.6, please use
Goo-Canvas version 0.03.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  perl >= 5.8.0
  Cairo >= 1.00 (perl module and requisite C libraries)
  Glib >= 1.103 (perl module and requisite C libraries)
  Gtk2 >= 1.100 (perl module and requisite C libraries)
  goocanvas-0.9


COPYRIGHT AND LICENCE

Copyright (C) 2007 by Ye Wenbin

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Goo::Canvas.

0.01  Wed Sep 26 16:24:09 2007
	- original version; created by h2xs 1.23 with options
		-A -n Goo::Canvas

0.03  Sun Oct  7 22:35:33 2007
	- Fix for demo/demo.pl create_stipples

0.04  Sun Oct  7 22:36:28 2007
    - Bindings for goocanvas-0.9

0.05  Sun Nov  4 14:25:34 2007
    - Fix get_bounds error
    - Add bin/perltetris.pl
    - Add bin/perlmine.pl

goocanvas-perl.h
/* @(#)goocanvas-perl.h
 */

#ifndef _GOOCANVAS_PERL_H
#define _GOOCANVAS_PERL_H 1

#include "gperl.h"
#include "gtk2perl.h"
#include "cairo-perl.h"
#include "goocanvas.h"

/* FIXME: How to avoid compile error in autogen.h */
typedef struct _GooCairoPattern GooCairoPattern;
struct _GooCairoPattern {
};
typedef struct _GooCairoMatrix GooCairoMatrix;
struct _GooCairoMatrix {
};

#include "goocanvas-perl-version.h"
#include "goocanvas-perl-autogen.h"

#define GOOCANVAS_PERL_VALUE_GET(pspec, value)
       

#define GOOCANVAS_PERL_ADD_PROPETIES(narg)                            \
    {                                                                 \
        GValue value = {0, };                                         \
        int i;                                                        \
        if ( 0 != (items-(narg)) % 2 )                                \
            croak ("set method expects name => value pairs "          \
		       "(odd number of arguments detected)");                 \
        for (i = (narg); i < items; i += 2) {                         \
            char *name = SvPV_nolen (ST (i));                         \
            SV *newval = ST (i + 1);                                  \
            GParamSpec *pspec;                                        \
            pspec = g_object_class_find_property(G_OBJECT_GET_CLASS(G_OBJECT(RETVAL)), name); \
            if ( !pspec ) {                                           \
                const char * classname =                              \
                    gperl_object_package_from_type(G_OBJECT_TYPE (G_OBJECT(RETVAL))); \
                if (!classname)                                       \
                    classname = G_OBJECT_TYPE_NAME(G_OBJECT(RETVAL)); \
                croak ("type %s does not support property '%s'",      \
                       classname, name);                              \
            }                                                         \
            g_value_init (&value, G_PARAM_SPEC_VALUE_TYPE (pspec));   \
            gperl_value_from_sv (&value, newval);                     \
            g_object_set_property (G_OBJECT(RETVAL), name, &value);   \
            g_value_unset (&value);                                   \
        }                                                             \
    }
    

#endif /* _GOOCANVAS_PERL_H */



lib/Goo/Canvas.pm
package Goo::Canvas;
use Gtk2;

use 5.008008;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Goo::Canvas ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.05';

require XSLoader;
XSLoader::load('Goo::Canvas', $VERSION);

# FIXME: Why ancestor not added?
push @ISA, 'Gtk2::Container';

# Preloaded methods go here.

1;
__END__
# documents.

=head1 NAME

Goo::Canvas - Perl interface to the GooCanvas

=head1 SYNOPSIS

    use Goo::Canvas;
    use Gtk2 '-init';
    use Glib qw(TRUE FALSE);

    my $window = Gtk2::Window->new('toplevel');
    $window->signal_connect('delete_event' => sub { Gtk2->main_quit; });
    $window->set_default_size(640, 600);

    my $swin = Gtk2::ScrolledWindow->new;
    $swin->set_shadow_type('in');
    $window->add($swin);

    my $canvas = Goo::Canvas->new();
    $canvas->set_size_request(600, 450);
    $canvas->set_bounds(0, 0, 1000, 1000);
    $swin->add($canvas);

    my $root = $canvas->get_root_item();
    my $rect = Goo::Canvas::Rect->new(
        $root, 100, 100, 400, 400,
        'line-width' => 10,
        'radius-x' => 20,
        'radius-y' => 10,
        'stroke-color' => 'yellow',
        'fill-color' => 'red'
    );
    $rect->signal_connect('button-press-event',
                          \&on_rect_button_press);

    my $text = Goo::Canvas::Text->new(
        $root, "Hello World", 300, 300, -1, 'center',
        'font' => 'Sans 24',
    );
    $text->rotate(45, 300, 300);
    $window->show_all();
    Gtk2->main;

    sub on_rect_button_press {
        print "Rect item pressed!\n";
        return TRUE;
    }

=head1 DESCRIPTION

GTK+ does't has an buildin canvas widget. GooCanvas is wonderful. It
is easy to use and has powerful and extensible way to create items in
canvas. Just try it.

For more documents, please read GooCanvas Manual and the demo programs
provided in the source distribution in both perl-Goo::Canvas and
GooCanvas.

=head1 SEE ALSO

L<Gtk2>(3pm)

=head1 AUTHOR

Ye Wenbin E<lt>wenbinye@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by ywb

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut


demo/scalablity.pl
#!/usr/bin/perl -w
# scalablity.pl --- 
# Last modify Time-stamp: <Ye Wenbin 2007-09-28 15:33:26>
# Version: v 0.0 2007/09/26 15:32:25
# Author: Ye Wenbin <wenbinye@gmail.com>

use strict;
use warnings;

use FindBin qw/$Bin/;
use lib "$Bin/../blib/arch";
use lib "$Bin/../blib/lib";

use Goo::Canvas;
use Gtk2 '-init';
use Glib qw(TRUE FALSE);
use Time::HiRes qw/gettimeofday tv_interval/;
use constant {
    N_GROUP_COLS => 5,
    N_GROUP_ROWS => 5,
    # N_GROUP_COLS => 25,
    # N_GROUP_ROWS => 20,
    N_COLS       => 10,
    N_ROWS       => 10,
    PADDING      => 10,
};
use Data::Dumper qw(Dumper); 
my $use_pixmap = 0;
my $max = 1<<29;
my ($left_offset, $top_offset, $total_width, $total_height);
my @ids;

my $window = Gtk2::Window->new('toplevel');
$window->signal_connect('delete_event' => sub { Gtk2->main_quit; });
$window->set_default_size(640, 600);

my $swin = Gtk2::ScrolledWindow->new;
$window->add($swin);

my $canvas = create_canvas();
$swin->add($canvas);

$window->show_all();
Gtk2->main();

sub create_canvas {
    my $canvas = Goo::Canvas->new();
    $canvas->set_size_request(600, 450);
    my $start = [gettimeofday];
    setup_canvas ($canvas);
    print "Create Canvas Time Used: ", tv_interval($start), "\n";
    $canvas->set_bounds($left_offset, $top_offset,
                        $left_offset + $total_width, $top_offset + $total_height);
    return $canvas;
}

sub setup_canvas {
    my $canvas = shift;
    my ($root, $group, $item);
    my ($pattern, $pixbuf);
    my ($item_width, $item_height, $group_width, $group_height, $cell_width, $cell_height);
    my @styles;
    my ($total_items, $id_item_num) = (0, 0);
    $root = $canvas->get_root_item();
    if ( $use_pixmap ) {
        die "Not support pixmap!";
        # $pixbuf = Gtk2::Gdk::Pixbuf->new_from_file("toriod.png");
        # $item_width = $pixbuf->get_width;
        # $item_height = $pixbuf->get_height;
        # $pattern = $pixbuf->
    } else {
        $item_width = 400;
        $item_height = 19;
    }
    $cell_width = $item_width + PADDING * 2;
    $cell_height = $item_height + PADDING * 2;
    $group_width = N_COLS * $cell_width;
    $group_height = N_ROWS * $cell_height;
    $total_width = N_GROUP_COLS * $group_width;
    $total_height = N_GROUP_ROWS * $group_height;
    $left_offset = - $total_width / 2;
    $top_offset = - $total_height / 2;
    for ( 'mediumseagreen', 'steelblue' ) {
        my $style = Goo::Canvas::Style->new;
        my $color = Gtk2::Gdk::Color->parse($_);
        my $pattern = Goo::Cairo::Pattern->new(
            Cairo::SolidPattern->create_rgb(
                map { $_/65535 } $color->red, $color->green, $color->blue
            ));
        $style->set_property('fill-pattern', $pattern);
        push @styles, $style;
    }
 OUTER:
    foreach my $i ( 0..N_GROUP_COLS ) {
        foreach my $j ( 0..N_GROUP_ROWS ) {
            my $x = $left_offset + ($i * $group_width);
            my $y = $top_offset + ($j * $group_height);
            $group = Goo::Canvas::Group->new( $root );
            $total_items++;
            $group->translate($x, $y);
            for my $i ( 0..N_COLS ) {
                for my $j ( 0..N_ROWS ) {
                    my $ix = $i * $cell_width + PADDING;
                    my $iy = $j * $cell_height + PADDING;
                    my $rotation = $i % 10 * 2;
                    my $rx = $ix + $item_width / 2;
                    my $ry = $iy + $item_height / 2;
                    $ids[$id_item_num] = ($x+$ix) . " - " . ($y+$iy);
                    $item = Goo::Canvas::Rect->new($group, $ix, $iy, $item_width, $item_height);
                    $item->set_style($styles[($j+1)%2]);
                    $item->rotate($rotation, $rx, $ry);
                    $item->{"id"} = $ids[$id_item_num];
                    $item->signal_connect('motion-notify-event',
                                          \&on_motion_notify);
                    $item = Goo::Canvas::Text->new($group,
                                                   $ids[$id_item_num],
                                                   $ix+$item_width/2, $iy+$item_height/2,
                                                   -1, 'center',
                                                   "font" => 'Sans 8');
                    $item->rotate($rotation, $rx, $ry);
                    $id_item_num++;
                    $total_items+=2;
                    if ( $max < $total_items ) {
                        last OUTER;
                    }
                }
            }
        }
    }
    print("total items: ", $total_items, "\n");
}
                                               
sub on_motion_notify {
    my $item = shift;
    print( ($item->{id} || "Unknown"),
        " item received 'motion-notify' signal\n");
    return FALSE;
}

    


Makefile.PL
#! /usr/bin/perl -w
use 5.008;
use strict;
use warnings;
use Cwd;
use File::Spec;
use ExtUtils::MakeMaker;

# minimum required version of dependancies we need to build
our %build_reqs = (
	'perl-ExtUtils-Depends'   => '0.2',
	'perl-ExtUtils-PkgConfig' => '1.0',
	'perl-Glib'               => '1.103',
	'perl-Gtk2'               => '1.100',
	'perl-Cairo'              => '1.00',
	'goocanvas'           => '0.9',
);

our %prereqs = (
	'Glib'                => $build_reqs{'perl-Glib'},
	'Gtk2'                => $build_reqs{'perl-Gtk2'},
	'Cairo'               => $build_reqs{'perl-Cairo'},
	'ExtUtils::Depends'   => $build_reqs{'perl-ExtUtils-Depends'},
	'ExtUtils::PkgConfig' => $build_reqs{'perl-ExtUtils-PkgConfig'},
);

# Writing a fake Makefile ensures that CPAN will pick up the correct
# dependencies and install them.
unless (eval "use ExtUtils::Depends;"
           . "use ExtUtils::PkgConfig;"
           . "use Gtk2::CodeGen;"
           . "use Cairo;"
           # just seeing if Glib is available isn't enough, make sure
           # it's recent enough, too 
           . "use Glib '$build_reqs{'perl-Glib'}';"
           . "use Gtk2 '$build_reqs{'perl-Gtk2'}';"
           . "use Glib::MakeHelper;"
           . "1") {
   warn "$@\n";
   WriteMakefile(
         PREREQ_FATAL => 1,
         PREREQ_PM    => \%prereqs,
   );
   exit 1; # not reached
}

mkdir 'build', 0777;

our %pkgcfg = ExtUtils::PkgConfig->find ('goocanvas');

# now we're ready to start creating the makefile.
# we need to use ExtUtils::Depends to get relevant information out of
# the Glib extension, and to save config information for other modules which
# will chain from this one.

our @xs_files = <xs/*.xs>;
our %pm_files = (
    'lib/Goo/Canvas.pm' => '$(INST_LIBDIR)/Canvas.pm',
);
our %pod_files;
%pod_files = (
    'lib/Goo/Canvas.pm' => '$(INST_MAN3DIR)/Goo::Canvas.$(MAN3EXT)',
    Glib::MakeHelper->do_pod_files (@xs_files),
);

ExtUtils::PkgConfig->write_version_macros (
    "build/goocanvas-perl-version.h",
    'goocanvas' => 'GOO_CANVAS'
);

#
# autogeneration
#
Gtk2::CodeGen->parse_maps ('goocanvas-perl');
Gtk2::CodeGen->write_boot (ignore => '^Goo::Canvas$');

my $goo = ExtUtils::Depends->new ('Goo::Canvas', 'Gtk2', 'Glib', 'Cairo');
$goo->set_inc ($pkgcfg{cflags} . ' -I./build ');
$goo->set_libs ($pkgcfg{libs});
$goo->add_xs (@xs_files);
$goo->add_pm (%pm_files);
my $cwd = cwd();
$goo->add_typemaps (
    File::Spec->catfile( $cwd, 'build/goocanvas-perl.typemap'),
    File::Spec->catfile( $cwd, 'goocanvas.typemap'),
);
$goo->install (qw(goocanvas-perl.h
		build/goocanvas-perl-autogen.h
		build/goocanvas-perl-version.h));
$goo->save_config ('build/IFiles.pm');
use Data::Dumper qw(Dumper);

WriteMakefile(
    NAME          => 'Goo::Canvas',
    VERSION_FROM  => 'lib/Goo/Canvas.pm', # finds $VERSION
    PREREQ_PM     => \%prereqs,
    ABSTRACT_FROM => 'lib/Goo/Canvas.pm', # retrieve abstract from module
    XSPROTOARG    => '-noprototypes',
    MAN3PODS	  => \%pod_files,
    $goo->get_makefile_vars,
);

print <<__EOW__;
WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING

   This is an unstable development release of Gtk2::SourceView.
   The API is not frozen and things are subject to change at any
   time.  Report any bugs to gtk-perl-list AT gnome DOT org as
   soon as possible. Please use the 1.00x series for important work.

WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
__EOW__

sub MY::postamble {
	return Glib::MakeHelper->postamble_clean ()
        . Glib::MakeHelper->postamble_docs (@main::xs_files)
            . Glib::MakeHelper->postamble_rpms (
                'GOO_CANVAS' => $build_reqs{'goocanvas'},
                'PERL_EXTUTILS_DEPENDS' => 
                    $build_reqs{'perl-ExtUtils-Depends'},
                'PERL_EXTUTILS_PKGCONFIG' => 
                    $build_reqs{'perl-ExtUtils-PkgConfig'},
                'PERL_GLIB' => $build_reqs{'perl-Glib'},
                'PERL_GTK' => $build_reqs{'perl-Gtk2'},
                'PERL_CAIRO' => $build_reqs{'perl-Cairo'},
            );
}
